### ISSUE TITLE
Allow generic typing of server & client methods

### ISSUE THREAD

**Issue by segevfiner** [shogowada/json-rpc-2.0#40]
It would be nice to be able to specify the types of server & client methods, to make sure that you call them correctly and haven't forgotten to implement any. Based on [`post-me`](https://github.com/alesgenova/post-me/blob/main/packages/core/src/handles.ts), I tried the following:
```ts
import { JSONRPCClient } from 'json-rpc-2.0';

type InnerType<T extends any | Promise<any>> = T extends Promise<
  infer U
>
  ? U
  : T;

type MethodsType = Record<string, (...args: any) => any>;

class MyJSONRPCClient<Methods extends MethodsType = any, ClientParams = void> extends JSONRPCClient<ClientParams> {
  request<K extends Extract<keyof Methods, string>>(
    method: K,
    params: Parameters<Methods[K]>,
    clientParams?: ClientParams
  ): PromiseLike<InnerType<ReturnType<Methods[K]>>> {
    return super.request(method, params, clientParams);
  }
}

type Foo = {
  foo(bla: number): number;
  bar(arg: string): string;
};

const x: MyJSONRPCClient<Foo>;
x.request('foo', [123]);

const y: MyJSONRPCClient;
y.request('bla', []);
```

Which works, except that `params` is now required, making it conditionally optional, is kind of difficult, I think it requires some hack with rest parameters and conditional types.

It would be nice to have this as a builtin feature of the library for both the server and client. Of course, this might be a breaking change, due to generic parameters order and should we not be able to make the default case type check as of today. Also we need to support an object as params, which might require more typescript trickery, such as an overload or more conditional types.

**Comment by segevfiner** [comment]
Here's a more complete version:
```ts
import { JSONRPCClient } from 'json-rpc-2.0';

// eslint-disable-next-line @typescript-eslint/no-explicit-any
type InnerType<T extends any | Promise<any>> = T extends Promise<infer U> ? U : T;

// eslint-disable-next-line @typescript-eslint/no-explicit-any
type MethodsType = Record<string, (params?: any) => any>;

class TypedJSONRPCClient<
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  Methods extends MethodsType = Record<string, (params?: any) => any>,
  ClientParams = void
> extends JSONRPCClient<ClientParams> {
  request<K extends Extract<keyof Methods, string>>(
    method: K,
    ...args: Parameters<Methods[K]>[0] extends undefined
      ? [params?: void, clientParams?: ClientParams]
      : [params: Parameters<Methods[K]>[0], clientParams?: ClientParams]
  ): PromiseLike<InnerType<ReturnType<Methods[K]>>>;
  request<K extends Extract<keyof Methods, string>>(
    method: K,
    params?: Parameters<Methods[K]>[0],
    clientParams?: ClientParams
  ): PromiseLike<InnerType<ReturnType<Methods[K]>>> {
    return super.request(method, params, clientParams);
  }
}

type Foo = {
  foo(params: [x: number]): number;
  bar(params: { foo: string }): string;
  noParams(): void;
};

const typed = new TypedJSONRPCClient<Foo>(async () => {
  return Promise.resolve();
});
typed.request('foo', [123]);
typed.request('bar', { foo: '123' });
// typed.request('bar'); // Error missing parameter
typed.request('noParams');
// typed.request('noParams', [123]); // Error passing argument to no argument

const untyped = new TypedJSONRPCClient(async () => {
  return Promise.resolve();
});
untyped.request('foo', [456]);
untyped.request('bar', { foo: 'hello' });
untyped.request('bar');
untyped.request('noParams');
untyped.request('noParams', [123]);
```

This just leaves whether we want a breaking change with `Methods` as the first generic parameter, or place it in the less logical second position to keep backwards compat.

We should also do something similar to the `JSONRPCServer`, though we might also want some exhaustiveness check there, but I'm not sure this can be achieved as TS has no static assert ATM, and it also can't be a runtime check as you can't turn types into actual values.

**Comment by segevfiner** [comment]
And for the server:
```ts
export class TypedJSONRPCServer<
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  Methods extends MethodsType = MethodsType,
  ServerParams = void
> extends JSONRPCServer<ServerParams> {
  addMethod<K extends Extract<keyof Methods, string>>(
    name: K,
    method: (
      params: Partial<Parameters<Methods[K]>[0]> | undefined,
      serverParams: ServerParams | undefined
    ) => ReturnType<Methods[K]> | PromiseLike<ReturnType<Methods[K]>>
  ): void {
    super.addMethod(name, method);
  }
}
```

**Comment by shogowada** [comment]
I'm wondering if it's worth creating `json-rpc-zod` or something that's based on `zod` implementation. I'm hesitant in doing it in this library because it adds a dependency, and it's TypeScript specific.

**Comment by shogowada** [comment]
@segevfiner It is now published under v1.5.0. I decided to only provide interfaces. Thank you for the idea!
